lib/collection/src/collection/shard_transfer.rs

```diff
@@ -340,7 +369,9 @@ impl Collection {
                 let shard_transfer_registered = shards_holder.shard_transfers.wait_for(
                     |shard_transfers| {
                         shard_transfers.iter().any(|shard_transfer| {
-                            shard_transfer.shard_id == shard_id && shard_transfer.to == this_peer_id
+                            (shard_transfer.shard_id == shard_id
+                                || shard_transfer.to_shard_id == Some(shard_id))
+                                && shard_transfer.to == this_peer_id
```

Comment:
I think we also have to explicitly check for `st.to_shard_id.is_none()`, e.g.:

```rust
(st.shard_id == shard_id && st.to_shard_id.is_none()) || st.to_shard_id == Some(shard_id)
```

Cleaned-up suggestion:

```suggestion
                            let to_this_shard = match shard_transfer.to_shard_id {
                                Some(to_shard_id) => to_shard_id == shard_id,
                                None => shard_transfer.shard_id == shard_id,
                            };

                            let to_this_peer = shard_transfer.to == this_peer_id;

                            to_this_shard && to_this_peer
```